traffic_ops/traffic_ops_golang/deliveryservice/acme.go

```diff
@@ -0,0 +1,363 @@
+	"database/sql"
+	"encoding/pem"
+	"errors"
+	"github.com/go-acme/lego/certcrypto"
+	"github.com/go-acme/lego/challenge"
+	"github.com/go-acme/lego/lego"
+	"github.com/go-acme/lego/registration"
+	"net/http"
+	"strconv"
+
+	"github.com/apache/trafficcontrol/lib/go-log"
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/auth"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/config"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/riaksvc"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/tenant"
+
+	"github.com/go-acme/lego/certificate"
+	"github.com/jmoiron/sqlx"
+)
+
+func RenewAcmeCertificate(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, []string{"xmlid"}, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+	if inf.Config.RiakEnabled == false {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, userErr, errors.New("deliveryservice.DeleteSSLKeys: Riak is not configured"))
+		return
+	}
+	xmlID := inf.Params["xmlid"]
+
+	if userErr, sysErr, errCode := tenant.Check(inf.User, xmlID, inf.Tx.Tx); userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+
+	ctx, _ := context.WithTimeout(r.Context(), LetsEncryptTimeout)
+
+	err := renewAcmeCerts(inf.Config, xmlID, ctx, inf.User)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, err)
+	}
+
+	api.WriteRespAlert(w, r, tc.SuccessLevel, "Certificate for "+xmlID+" successfully renewed.")
+
+}
+
+func renewAcmeCerts(cfg *config.Config, dsName string, ctx context.Context, currentUser *auth.CurrentUser) error {
+	db, err := api.GetDB(ctx)
+	if err != nil {
+		log.Errorf(dsName+": Error getting db: %s", err.Error())
+		return err
+	}
+
+	tx, err := db.Begin()
+	if err != nil {
+		log.Errorf(dsName+": Error getting tx: %s", err.Error())
+		return err
+	}
+
+	userTx, err := db.Begin()
+	if err != nil {
+		log.Errorf(dsName+": Error getting userTx: %s", err.Error())
+		return err
+	}
+	defer userTx.Commit()
+
+	logTx, err := db.Begin()
+	if err != nil {
+		log.Errorf(dsName+": Error getting logTx: %s", err.Error())
+		return err
+	}
+	defer logTx.Commit()
+
+	dsID, certVersion, err := getDSIdAndVersionFromName(db, dsName)
+	if err != nil {
+		return errors.New("querying DS info: " + err.Error())
+	}
+	if dsID == nil || *dsID == 0 {
+		return errors.New("DS id for " + dsName + " was nil or 0")
+	}
+	if certVersion == nil || *certVersion == 0 {
+		return errors.New("certificate for " + dsName + " could not be renewed because version was nil or 0")
+	}
+
+	keyObj, ok, err := riaksvc.GetDeliveryServiceSSLKeysObjV15(dsName, strconv.Itoa(int(*certVersion)), tx, cfg.RiakAuthOptions, cfg.RiakPort)
```

Comment:
Could you add a quick nil check for `cfg` before dereferencing it here?

Reply:
i dont think cfg could be nil without an error when doing api.NewInfo, but i threw in a nil check anyway